src/fsharp/PrettyNaming.fs

```diff
@@ -109,66 +109,160 @@ module internal Microsoft.FSharp.Compiler.PrettyNaming
+                let opName = sb.ToString ()
+
+                // Cache the compiled name so it can be reused.
+                lock compiledOperators <| fun () ->
+                    compiledOperators.[op] <- opName
+
+                opName
+
+    // +++ GLOBAL STATE
+    /// Compiles a custom operator into a mangled operator name.
+    /// For example, "!%" becomes "op_DereferencePercent"
     let CompileOpName =
-        let t = Map.ofList opNameTable
-        let t2 = Map.ofList opCharTranslateTable
-        fun n -> 
-            match t.TryFind(n) with 
-            | Some(x) -> x 
-            | None -> 
-                if IsOpName n then 
-                  let mutable r = []
-                  for i = 0 to String.length n - 1 do
-                      let c = n.[i]
-                      let c2 = match t2.TryFind(c) with Some(x) -> x | None -> string c
-                      r <- c2 :: r 
-                  "op_"^(String.concat "" (List.rev r))
-                else n
-
-    let IsMangledOpName (n:string) = n.Length >= 3 && n.Substring(0,3) = "op_"     
-                             
-    let DecompileOpName = 
-      let t = new Dictionary<string,string>()
-      for (x,y) in opNameTable do
-          t.Add(y,x)
-      fun n -> 
-          let mutable res = Unchecked.defaultof<_>
-          if t.TryGetValue(n,&res) then 
-              res
-          else
-              if n.StartsWith("op_",System.StringComparison.Ordinal) then 
-                let rec loop (remaining:string) = 
+        /// Maps the built-in F# operators to their mangled operator names.
+        let standardOpNames =
+            let opNames = Dictionary<_,_> (opNameTable.Length, System.StringComparer.Ordinal)
+            for x, y in opNameTable do
+                opNames.Add (x, y)
+            opNames
+
+        fun op ->
+            match standardOpNames.TryGetValue op with
+            | true, x -> x
+            | false, _ ->
+                if IsOpName op then
+                    compileCustomOpName op
+                else op
+
+    // +++ GLOBAL STATE
+    /// Decompiles the mangled name of a custom operator back into an operator.
+    /// For example, "op_DereferencePercent" becomes "!%"
+    let private decompileCustomOpName =
+        // Memoize this operation. Custom operators are typically used more than once
+        // so this avoids repeating decompilation.
+        let decompiledOperators = Dictionary<_,_> (System.StringComparer.Ordinal)
+
+        fun opName ->
+            // Has this operator name already been decompiled?
+            let found =
+                lock decompiledOperators <| fun () ->
+                    match decompiledOperators.TryGetValue opName with
+                    | true, res -> Some res
+                    | false, _ -> None
+
+            match found with
+            | Some res -> res
+            | None ->
+                let rec loop (remaining : string) =
```

Comment:
If you are looking to micro-optimize, this loop could return unit, and pass along a StringBuilder which is built up with the decompiled name.  That would make it tail-recursive (compiler will optimize into imperative loop), avoid some of the `Some` allocations, and avoid string concats.